packages/controllers/src/fsm.ts

```diff
@@ -0,0 +1,96 @@
+import {
+  EventObject,
+  InterpreterStatus,
+  StateMachine,
+  Typestate,
+} from '@xstate/fsm';
+import { assert } from './utils';
+
+/**
+ * Validates the set-up of a @xstate/fsm machine.
+ *
+ * 1. Ensures that all named actions in the config have a provided implementation.
+ *
+ * @param machine - The machine to validate.
+ * @throws {@link AssertionError}. If the validation fails.
+ */
+export function validateMachine<
+  TContext extends object,
+  TEvent extends EventObject,
+  TState extends Typestate<TContext>,
+>(machine: StateMachine.Machine<TContext, TEvent, TState>) {
+  assert('_options' in machine, 'The machine is not an @xstate/fsm machine');
+  const typed = machine as StateMachine.Machine<TContext, TEvent, TState> & {
+    _options: { actions?: StateMachine.ActionMap<TContext, TEvent> };
+  };
+
+  // 1.
+  const toArray = <T>(obj: T | T[]): T[] => {
+    if (Array.isArray(obj)) {
+      return obj;
+    } else if (obj === undefined || obj === null) {
+      return [];
+    }
+    return [obj];
+  };
+  const allActions = new Set<string>();
+  const addActions = (actions: any) =>
+    toArray(actions)
+      .flatMap<string>((action) => {
+        if (typeof action === 'string') {
+          return [action];
+        }
+        assert(typeof action === 'function');
+        return [];
+      })
+      .forEach(allActions.add.bind(allActions));
+
+  for (const state of Object.values<typeof typed.config.states[string]>(
+    typed.config.states,
+  )) {
+    addActions(state.entry);
+    addActions(state.exit);
+    for (const transition of Object.values<any>(state.on ?? {})) {
+      addActions(transition.actions);
+    }
+  }
+
+  allActions.forEach((action) =>
+    assert(
+      typed._options.actions !== undefined && action in typed._options.actions,
+      `Action "${action}" doesn't have an implementation`,
+    ),
+  );
+}
```

Comment:
Can you expand on what this does and why is it needed? How / where do we specify implementations for actions?

Reply:
Ah, I see. We don't have any XState actions implemented right now, but we will in the future? How soon do you anticipate that we will use actions?

Reply:
I've used actions extensively in #658, moved this function from there. It's up to us when we want to use them, we could start small, for example publishing to messenger on state entry.